File: Ejercicios Resueltos/C++/ejercicio5.cpp
#include <iostream>
#include <string>
#include <queue>

using namespace std;

struct NodoWeb {
    string modulo;
    NodoWeb* izquierdo;
    NodoWeb* derecho;
};

NodoWeb* crearNodo(string nombre) {
    NodoWeb* nuevo = new NodoWeb();
    nuevo->modulo = nombre;
    nuevo->izquierdo = NULL;
    nuevo->derecho = NULL;
    return nuevo;
}

void mostrarMenuPrincipal(NodoWeb* raiz, string sangria = "") {
    if (raiz == NULL) return;
    cout << sangria << "- " << raiz->modulo << endl;
    mostrarMenuPrincipal(raiz->izquierdo, sangria + "   ");
    mostrarMenuPrincipal(raiz->derecho, sangria + "   ");
}

void procesarModulosInternos(NodoWeb* raiz) {
    if (raiz == NULL) return;
    procesarModulosInternos(raiz->izquierdo);
    procesarModulosInternos(raiz->derecho);
    cout << "Cargando dependencias de: [" << raiz->modulo << "]" << endl;
}

void mostrarNivelPorNivel(NodoWeb* raiz) {
    if (raiz == NULL) return;
    queue<NodoWeb*> cola;
    cola.push(raiz);
    int nivel = 0;
    while (!cola.empty()) {
        int nodosEnNivel = cola.size();
        cout << "Profundidad " << nivel << ": ";
        for (int i = 0; i < nodosEnNivel; i++) {
            NodoWeb* actual = cola.front();
            cola.pop();
            cout << actual->modulo << " | ";
            if (actual->izquierdo != NULL) cola.push(actual->izquierdo);
            if (actual->derecho != NULL) cola.push(actual->derecho);
        }
        cout << endl;
        nivel++;
    }
}

int main() {
    NodoWeb* raiz = crearNodo("Sistema Web");
    raiz->izquierdo = crearNodo("Usuarios");
    raiz->derecho = crearNodo("Inventario");
    raiz->izquierdo->izquierdo = crearNodo("Registrar");
    raiz->izquierdo->derecho = crearNodo("Buscar");
    raiz->derecho->izquierdo = crearNodo("Productos");
    raiz->derecho->derecho = crearNodo("Reportes");

    int opcion;

    do {
        cout << "\n=== MENU INTERACTIVO: SISTEMA WEB ===" << endl;
        cout << "1. Mostrar Menu Principal (Preorden)" << endl;
        cout << "2. Procesar Modulos Internos (Postorden)" << endl;
        cout << "3. Mostrar Mapa del Sitio (BFS)" << endl;
        cout << "4. Salir" << endl;
        cout << "Ingrese una opcion: ";
        cin >> opcion;

        cout << "\n-----------------------------------" << endl;
        switch(opcion) {
            case 1:
                cout << "=== MENU PRINCIPAL ===" << endl;
                mostrarMenuPrincipal(raiz);
                break;
            case 2:
                cout << "=== PROCESANDO MODULOS ===" << endl;
                procesarModulosInternos(raiz);
                break;
            case 3:
                cout << "=== MAPA DEL SITIO ===" << endl;
                mostrarNivelPorNivel(raiz);
                break;
            case 4:
                cout << "Saliendo del programa..." << endl;
                break;
            default:
                cout << "Opcion no valida. Intente de nuevo." << endl;
        }
        cout << "-----------------------------------" << endl;
        
    } while(opcion != 4);

    return 0;
}

File: Ejercicios Resueltos/Java/ejercicio5.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class NodoWeb {
    String modulo;
    NodoWeb izquierdo, derecho;

    public NodoWeb(String modulo) {
        this.modulo = modulo;
        this.izquierdo = null;
        this.derecho = null;
    }
}

public class ejercicio5 {

    public static void mostrarMenuPrincipal(NodoWeb raiz, String sangria) {
        if (raiz == null) return;
        System.out.println(sangria + "- " + raiz.modulo);
        mostrarMenuPrincipal(raiz.izquierdo, sangria + "   ");
        mostrarMenuPrincipal(raiz.derecho, sangria + "   ");
    }

    public static void procesarModulosInternos(NodoWeb raiz) {
        if (raiz == null) return;
        procesarModulosInternos(raiz.izquierdo);
        procesarModulosInternos(raiz.derecho);
        System.out.println("Cargando dependencias de: [" + raiz.modulo + "]");
    }

    public static void mostrarNivelPorNivel(NodoWeb raiz) {
        if (raiz == null) return;
        Queue<NodoWeb> cola = new LinkedList<>();
        cola.add(raiz);
        int nivel = 0;
        while (!cola.isEmpty()) {
            int nodosEnNivel = cola.size();
            System.out.print("Profundidad " + nivel + ": ");
            for (int i = 0; i < nodosEnNivel; i++) {
                NodoWeb actual = cola.poll();
                System.out.print(actual.modulo + " | ");
                if (actual.izquierdo != null) cola.add(actual.izquierdo);
                if (actual.derecho != null) cola.add(actual.derecho);
            }
            System.out.println();
            nivel++;
        }
    }

    public static void main(String[] args) {
        // Construcción manual del árbol del Sistema Web
        NodoWeb raiz = new NodoWeb("Sistema Web");
        raiz.izquierdo = new NodoWeb("Usuarios");
        raiz.derecho = new NodoWeb("Inventario");
        raiz.izquierdo.izquierdo = new NodoWeb("Registrar");
        raiz.izquierdo.derecho = new NodoWeb("Buscar");
        raiz.derecho.izquierdo = new NodoWeb("Productos");
        raiz.derecho.derecho = new NodoWeb("Reportes");

        Scanner scanner = new Scanner(System.in);
        int opcion;

        do {
            System.out.println("\n=== MENU INTERACTIVO: SISTEMA WEB ===");
            System.out.println("1. Mostrar Menu Principal (Preorden)");
            System.out.println("2. Procesar Modulos Internos (Postorden)");
            System.out.println("3. Mostrar Mapa del Sitio (BFS)");
            System.out.println("4. Salir");
            System.out.print("Ingrese una opcion: ");
            
            opcion = scanner.nextInt();

            System.out.println("\n-----------------------------------");
            switch(opcion) {
                case 1:
                    System.out.println("=== MENU PRINCIPAL ===");
                    mostrarMenuPrincipal(raiz, "");
                    break;
                case 2:
                    System.out.println("=== PROCESANDO MODULOS ===");
                    procesarModulosInternos(raiz);
                    break;
                case 3:
                    System.out.println("=== MAPA DEL SITIO ===");
                    mostrarNivelPorNivel(raiz);
                    break;
                case 4:
                    System.out.println("Saliendo del programa...");
                    break;
                default:
                    System.out.println("Opcion no valida. Intente de nuevo.");
            }
            System.out.println("-----------------------------------");
            
        } while(opcion != 4);

        scanner.close();
    }
}